docs/concepts/components/properties.md

```diff
@@ -71,3 +94,21 @@ pub struct LinkProps {
 }
 ```
 
+## Props macro
+
+The `yew::props!` macro allows you to build properties outside of the `html!` macro but with the same syntax.
+
+The macro invocation starts with the props type.
+You can either point to the props directly (`path::to::Props`) or use the associated properties of a component (`MyComp::Properties`).
+
+```rust
+let props = yew::props!(LinkProps
+    href="/"
+    text=Rc::from("imagine this text being really long")
+    size=64
+);
```

Comment:
I'm wondering if it would make sense to mirror the struct syntax instead. I think the `prop=value` syntax in particular is a little odd outside of macro. What about this:
```suggestion
let props = yew::props!(LinkProps {
    href: "/",
    text: Rc::from("imagine this text being really long"),
    size: 64,
});
```

Reply:
I see where you're coming from. When I wrote the documentation and had to split this it across multiple lines for readability my first thought was "huh, that doesn't look as good as I was expecting".

Here are a few points in favour of the current syntax that come to mind:

- because it's the same syntax you can move code from `html!` to `yew::props!` or vice versa and it still works.
- it's more obvious that the same rules apply, especially with things like transformers. I feel like the fake-Rust syntax can easily lead to wrong assumptions.
- the `yew::props!` macro uses the exact same parser as the props in the `html!` macro. This makes it much easier for the two to stay in sync. This also makes it easier for a potential Yew IDE extension.
- if we want to add special syntax in the future (like `?=`) it would be very hard to reproduce this in the `yew::props!` macro without breaking the Rust syntax anyway.


```rust
html! {
  <Foo
    title_props=yew::props!(Title::Properties text="hello world")
    body_props=yew::props!(Body::Properties
      class="color-red"
      onhover=self.link.callback(Msg::HoverBody)
    )
    onclick=self.link.callback(Msg::Click)
  />
}

// vs

html! {
  <Foo
    title_props=yew::props!(Title::Properties { text: "hello world" })
    body_props=yew::props!(Body::Properties {
      class: "color-red",
      onhover: self.link.callback(Msg::HoverBody)
    })
    onclick=self.link.callback(Msg::Click)
  />
}
```

Personally, I think the first one is less distracting.

Reply:
> because it's the same syntax you can move code from html! to yew::props! or vice versa and it still works.

Good point!

> it's more obvious that the same rules apply, especially with things like transformers. I feel like the fake-Rust syntax can easily lead to wrong assumptions.

Yeah fair point about the transformers but, to be honest, I think transformer behaviour is not usually that obvious anyways.

> the yew::props! macro uses the exact same parser as the props in the html! macro. This makes it much easier for the two to stay in sync. It should also make it easier for a potential Yew IDE extension.

They are actually already different in regards with "with props" and special props so maybe it actually makes sense for them to diverge a bit. The nice thing is that the tokenizer code wouldn't have to change.

> if we want to add special syntax in the future (like ?=) it would be very hard to reproduce this in the yew::props! macro without breaking the fake-Rust syntax.

This is probably the most compelling reason to me to keep the "label = value" syntax. Is it pretty likely we'll need to add that special syntax?

---

Some points in favour of struct syntax
- I imagine that rustfmt will help make props look nice in the struct form
- You can move code easily between using yew::props! and not using the macro
- when yew::props! is used separately from the JSX syntax it looks quite strange

Reply:
> This is probably the most compelling reason to me to keep the "label = value" syntax. Is it pretty likely we'll need to add that special syntax?

I'm not sure, I had two-way bindings in mind, but something like that only makes sense in the context of a component so it wouldn't work in `yew::props!` anyway.
For now I don't think there's any reason to worry about special syntax.

<br>

> I imagine that rustfmt will help make props look nice in the struct form

I'm fairly certain that doesn't work. As far as I can tell Rustfmt is entirely disabled in proc macros by default (Because of https://github.com/rust-lang/rustfmt/issues/3434).

<br>

> You can move code easily between using yew::props! and not using the macro

Moving away from the macro still requires the user to add the fields with default values and manually add the transformations performed by `Transformer`. It's certainly true the other way around though.

---

I came up with some reasons in favour of the struct-like syntax of my own:

- You can use the field init shorthand (`a` instead of `a: a`)
- In the future we could support the base expression syntax (`Props { text: "a", ..other_props }`)

Either syntax has its fair share of problems but I'm totally fine with going with the struct syntax.

Reply:
> I'm fairly certain that doesn't work. As far as I can tell Rustfmt is entirely disabled in proc macros by default (Because of rust-lang/rustfmt#3434).

Got it, makes sense.

> Moving away from the macro still requires the user to add the fields with default values and manually add the transformations performed by Transformer. It's certainly true the other way around though.

Ah yeah, I meant the other way around 👍 

> You can use the field init shorthand (a instead of a: a)

Good point!

> Either syntax has its fair share of problems but I'm totally fine with going with the struct syntax.

Certainly true. Sounds like struct syntax has the edge, shall we go with that then?

Reply:
Oh! You already made the change! Looks good!